### PR TITLE
create and export compartment ID capabilities

### PR DESCRIPTION
**Pull request by brooksdavis**
- Improve sysctl support for exporting capabilities
- Create a root CID capability on Morello
- (Temporarily) export the CID capability via the security.cheri.cidcap sysctl
- Add cheri_cidcap_alloc(2) syscall

### CODE REVIEW

File: sys/arm64/include/cherireg.h [hunk 1]
```diff
@@ -169,4 +175,17 @@
...
+ * Root compartment ID capablity for userspace.
+ *
+ * XXX: move these to sys/cheri/cherireg.h if another platform implements CIDs.
+ *
+ * XXX: a larger CID space is possible, but unr(9) only supports UINT_MAX
+ * so go with that for now.
+ */
+#define	CHERI_COMPARTMENT_ID_USERSPACE_PERMS	CHERI_PERMS_USERSPACE_CIDCAP
+#define	CHERI_COMPARTMENT_ID_USERSPACE_BASE	0x0
+#define	CHERI_COMPARTMENT_ID_USERSPACE_LENGTH	0xFFFFFFFF
```
Comment (brooksdavis) [inline sys/arm64/include/cherireg.h:188]: I'm going to change this to 0x8FFFFFFFFFFFFFFF and use unr64 to implement the allocator.

File: bin/cheribsdtest/Makefile.cheribsdtest [hunk 2]
```diff
@@ -8,6 +8,7 @@ SRCS+=	cheribsdtest_bounds_globals.c					\
 	cheribsdtest_bounds_globals_x.c					\
 	cheribsdtest_bounds_heap.c					\
 	cheribsdtest_bounds_stack.c					\
```
Comment (bsdjhb) [inline bin/cheribsdtest/Makefile.cheribsdtest:10]: Nit: s/This is a inspired/This is inspired/ in the commit log

File: bin/cheribsdtest/cheribsdtest_cidcap.c [hunk 3]
```diff
@@ -95,7 +98,20 @@ CHERIBSDTEST(cidcap_sysctl, "Retrieve cidcap using sysctl(3)")
...
+	cidcap = get_cidcap_sysctl();
+
+	check_cidcap(cidcap, CHERI_COMPARTMENT_ID_USERSPACE_BASE,
+	    CHERI_COMPARTMENT_ID_USERSPACE_LENGTH,
+	    CHERI_COMPARTMENT_ID_USERSPACE_OFFSET);
+
         cheribsdtest_success();
 }
 
+
```
Comment (bsdjhb) [inline bin/cheribsdtest/cheribsdtest_cidcap.c:139]: Stray newline?

File: bin/cheribsdtest/cheribsdtest_cidcap.c [hunk 4]
```diff
@@ -60,9 +60,9 @@ check_cidcap(uintcap_t cidcap, size_t base, size_t length, size_t offset)
 	if (base != (size_t)-1) {
 		/* Base. */
 		v = cheri_getbase(cidcap);
-		if (v != CHERI_COMPARTMENT_ID_USERSPACE_BASE)
```
Comment (bsdjhb) [inline bin/cheribsdtest/cheribsdtest_cidcap.c]: These fixes belong in the previous commit

File: sys/cheri/cheri_cidcap.c [hunk 5]
```diff
@@ -31,19 +31,61 @@
...
+	struct vm_map *map;
+	uint64_t cid;
+	uintcap_t cidcap;
+
+	map = &td->td_proc->p_vmspace->vm_map;
+	vm_map_lock(map);
+	if (map->busy)
+		vm_map_wait_busy(map);
+
+	cid = td->td_proc->p_vmspace->vm_next_cid++;
```
Comment (bsdjhb) [inline sys/cheri/cheri_cidcap.c:69]: Maybe at least KASSERT that we don't rollover?

File: sys/cheri/cidcap.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifndef __SYS_CHERI_CIDCAP_H__
+#define	__SYS_CHERI_CIDCAP_H__
+
+#ifndef _KERNEL
+/* XXX: currently Morello specific */
+int cheri_cidcap_alloc(uintcap_t *cidcap);
```
Comment (bsdjhb) [inline sys/cheri/cidcap.h:38]:
```suggestion
int	cheri_cidcap_alloc(uintcap_t *cidcap);
```

**Review by bsdjhb — APPROVED** [review]
One suggestion perhaps would be to more the "refactor sealcap tests" commit up one before the addition of the first "cidcap" test as the initial cidcap tests look to be more closely derived from sealcap after the refactoring.

**Review by bsdjhb — APPROVED** [review]

**Pull request merged**